packages/app-admin/src/components/IconPicker/IconRepository.ts

```diff
@@ -2,47 +2,72 @@ import cloneDeep from "lodash/cloneDeep";
-import {
-    IconPackProviderInterface as IconPackProvider,
-    IconType
-} from "~/components/IconPicker/config";
-import { ProviderIcon } from "~/components/IconPicker/config/IconPackProvider";
+import { CustomIconsGatewayInterface } from "./gateways";
+import { IconPackProviderInterface as IconPackProvider, IconType } from "./config";
+import { ProviderIcon } from "./config/IconPackProvider";
 
 export class IconRepository {
-    public readonly iconPackProviders: IconPackProvider[];
+    private gateway: CustomIconsGatewayInterface;
+    private readonly iconPackProviders: IconPackProvider[];
     private readonly iconTypes: IconType[];
     private loading: Loading;
     private icons: ProviderIcon[] = [];
+    private customIcons: ProviderIcon[] = [];
 
-    constructor(iconTypes: IconType[], iconPackProviders: IconPackProvider[]) {
+    constructor(
+        gateway: CustomIconsGatewayInterface,
+        iconTypes: IconType[],
+        iconPackProviders: IconPackProvider[]
+    ) {
+        this.gateway = gateway;
         this.iconTypes = iconTypes;
         this.loading = new Loading(true);
         this.iconPackProviders = iconPackProviders;
         makeAutoObservable(this);
     }
 
     async loadIcons() {
-        if (this.icons.length > 0) {
-            return;
+        if (this.icons.length === 0) {
+            const icons = await this.runWithLoading(async () => {
+                const icons = await Promise.all(
+                    this.iconPackProviders.map(provider => provider.getIcons())
+                );
+                return icons.flat();
+            });
+
+            const iconTypes = this.iconTypes.map(iconType => iconType.name);
+
+            runInAction(() => {
+                // Make sure we only work with known icon types.
+                this.icons = icons.filter(icon => iconTypes.includes(icon.type));
+            });
         }
 
-        const icons = await this.runWithLoading(async () => {
-            const icons = await Promise.all(
-                this.iconPackProviders.map(provider => provider.getIcons())
-            );
-            return icons.flat();
-        });
+        if (this.customIcons.length === 0) {
+            const customIcons = await this.runWithLoading(async () => {
+                return await this.gateway.listCustomIcons();
+            });
 
-        const iconTypes = this.iconTypes.map(iconType => iconType.name);
+            if (!customIcons) {
+                return;
+            }
 
-        runInAction(() => {
-            // Make sure we only work with known icon types.
-            this.icons = icons.filter(icon => iconTypes.includes(icon.type));
-        });
+            runInAction(() => {
+                this.customIcons = customIcons.map(customIcon => ({
+                    type: "custom",
+                    name: customIcon.name,
+                    value: customIcon.src
+                }));
+            });
+        }
```

Comment:
This whole thing should not exist in the Repository. It should be coming from an IconPackProvider. So `custom` icons are nothing more than an API call to fetch files from the File Manager. Repository already works with icon providers, so there's nothing for you to do in this file. And adding new dependencies, especially something like `Custom` should immediately ring the alarm.